src/Routers/UsersRouter.js

```diff
@@ -88,94 +68,154 @@ export class UsersRouter extends ClassesRouter {
+          }
 
-        return passwordCrypto.compare(password, user.password);
-      })
-      .then((correct) => {
-        isValidPassword = correct;
-        const accountLockoutPolicy = new AccountLockout(user, req.config);
-        return accountLockoutPolicy.handleLoginAttempt(isValidPassword);
-      })
-      .then(() => {
-        if (!isValidPassword) {
-          throw new Parse.Error(Parse.Error.OBJECT_NOT_FOUND, 'Invalid username/password.');
-        }
-        if (req.config.verifyUserEmails && req.config.preventLoginWithUnverifiedEmail && !user.emailVerified) {
-          throw new Parse.Error(Parse.Error.EMAIL_NOT_FOUND, 'User email is not verified.');
-        }
-        // handle password expiry policy
-        if (req.config.passwordPolicy && req.config.passwordPolicy.maxPasswordAge) {
-          let changedAt = user._password_changed_at;
-
-          if (!changedAt) {
-            // password was created before expiry policy was enabled.
-            // simply update _User object so that it will start enforcing from now
-            changedAt = new Date();
-            req.config.database.update('_User', {username: user.username},
-              {_password_changed_at: Parse._encode(changedAt)});
+          if (results.length > 1) { // corner case where user1 has username == user2 email
+            req.config.loggerController.warn('There is a user which email is the same as another user\'s username, logging in based on username');
+            user = results.filter((user) => user.username === username)[0];
           } else {
-            // check whether the password has expired
-            if (changedAt.__type == 'Date') {
-              changedAt = new Date(changedAt.iso);
+            user = results[0];
+          }
+
+          return passwordCrypto.compare(password, user.password);
+        })
+        .then((correct) => {
+          isValidPassword = correct;
+          const accountLockoutPolicy = new AccountLockout(user, req.config);
+          return accountLockoutPolicy.handleLoginAttempt(isValidPassword);
+        })
+        .then(() => {
+          if (!isValidPassword) {
+            throw new Parse.Error(Parse.Error.OBJECT_NOT_FOUND, 'Invalid username/password.');
+          }
+          if (req.config.verifyUserEmails && req.config.preventLoginWithUnverifiedEmail && !user.emailVerified) {
+            throw new Parse.Error(Parse.Error.EMAIL_NOT_FOUND, 'User email is not verified.');
+          }
+          // handle password expiry policy
+          if (req.config.passwordPolicy && req.config.passwordPolicy.maxPasswordAge) {
+            let changedAt = user._password_changed_at;
+
+            if (!changedAt) {
+              // password was created before expiry policy was enabled.
+              // simply update _User object so that it will start enforcing from now
+              changedAt = new Date();
+              req.config.database.update('_User', { username: user.username },
```

Comment:
this part feels odd that we keep it in a verification, when the password is actually not changed there. What do you think?

Reply:
Makes sense! I'll move this snippet back into the handleLogin function.

Reply:
Exellent thanks! That,s quite the refactor you're on, I want to get it right before we merge ;)